plugin/src/main/java/org/opensearch/ml/action/memorycontainer/memory/TransportAddMemoriesAction.java

```diff
@@ -0,0 +1,1607 @@
+            return null;
+        }
+
+        ModelTensors modelTensors = tensorOutput.getMlModelOutputs().get(0);
+        if (modelTensors.getMlModelTensors() == null || modelTensors.getMlModelTensors().isEmpty()) {
+            log.debug("No model tensors found in model output");
+            return null;
+        }
+
+        // For dense embeddings, look for the sentence_embedding tensor
+        for (ModelTensor tensor : modelTensors.getMlModelTensors()) {
+            if ("sentence_embedding".equals(tensor.getName()) && tensor.getData() != null) {
+                Number[] data = tensor.getData();
+                log.debug("Found sentence_embedding tensor with dimension: {}", data.length);
+
+                // Convert Number[] to float[] for proper storage
+                float[] floatData = new float[data.length];
+                for (int i = 0; i < data.length; i++) {
+                    floatData[i] = data[i].floatValue();
+                }
+                return floatData;
+            }
+        }
+
+        log.error("No sentence_embedding tensor found for dense embedding");
+        return null;
+    }
+
+    private Map<String, ?> buildSparseEmbeddingFromResponse(ModelTensorOutput tensorOutput) {
+        if (tensorOutput.getMlModelOutputs() == null || tensorOutput.getMlModelOutputs().isEmpty()) {
+            log.debug("No model outputs found in tensor output");
+            return null;
+        }
+
+        ModelTensors modelTensors = tensorOutput.getMlModelOutputs().get(0);
+        if (modelTensors.getMlModelTensors() == null || modelTensors.getMlModelTensors().isEmpty()) {
+            log.debug("No model tensors found in model output");
+            return null;
+        }
+
+        // For sparse embeddings, find the tensor with dataAsMap
+        for (ModelTensor tensor : modelTensors.getMlModelTensors()) {
+            Map<String, ?> dataMap = tensor.getDataAsMap();
+            if (dataMap != null) {
+                // Check if sparse embedding is nested in a response field
+                if (dataMap.containsKey("response") && dataMap.get("response") instanceof List) {
+                    List<?> responseList = (List<?>) dataMap.get("response");
+                    if (!responseList.isEmpty() && responseList.get(0) instanceof Map) {
+                        Map<String, ?> sparseMap = (Map<String, ?>) responseList.get(0);
+                        log.debug("Extracted sparse embedding from nested response with {} tokens", sparseMap.size());
+                        return sparseMap;
+                    }
+                }
+                // Otherwise return the direct map
+                log.debug("Using direct sparse embedding with {} tokens", dataMap.size());
+                return dataMap;
+            }
+        }
+
+        log.error("No sparse embedding data found");
+        return null;
+    }
+
+    private void searchSimilarFactsForSession(
```

Comment:
This part we might need to have more discussion (for 3.3). Do we want to search the facts only for the session? 

Facts are treated as a long term memory. A long term memory should orthogonal to a particular session I think. But for 3.2 I'm fine with this change. 

We should have more discussion on this later.